### PR TITLE
Allow ASCII-8BIT string for white list sanitize.

### PR DESCRIPTION
**Pull request by JunichiIto**
Fix this issue: https://github.com/rails/rails-html-sanitizer/issues/30

Also related to this issue: https://github.com/rails/rails/issues/18975

### CODE REVIEW

**Comment by JunichiIto** [comment]
Anyone??

File: lib/rails/html/sanitizer.rb [hunk 1]
```diff
@@ -106,7 +106,8 @@ def sanitize(html, options = {})
         return unless html
         return html if html.empty?
 
-        loofah_fragment = Loofah.fragment(html)
+        encoding = html.encoding == Encoding::ASCII_8BIT ? Encoding::UTF_8 : html.encoding
+        loofah_fragment = Loofah.fragment(html, encoding.name)
```
Comment (kaspth) [inline lib/rails/html/sanitizer.rb]: Let's do it this way:

```ruby
loofah_fragment = Loofah.fragment(html, pick_encoding(html))

# later in the private methods

def pick_encoding(html)
  'UTF-8' if html.encoding == Encoding::ASCII_8BIT # Nokogiri doesn't support ASCII-8BIT
end
```

**Comment by kaspth** [comment]
:+1:, but remember to squash your commits into one.

**Comment by JunichiIto** [comment]
@kaspth Thank you for review. I did refactoring and squashed.

**Comment by JunichiIto** [comment]
I'm looking forward to getting merged... :kissing_closed_eyes:

**Comment by kaspth** [comment]
Don't get cocky, kid :wink:

Especially given the concerns on the original issue here: https://github.com/rails/rails-html-sanitizer/issues/30#issuecomment-75860863. Instead we're going to warn about the mysql gem. Thanks.

**Comment by JunichiIto** [comment]
Wait, my main RDBMS is PostgreSQL, not MySQL. 

I got this result in my Rails 4.2 app without any dependencies on MySQL.

```ruby
irb(main):007:0> snp=ReferenceSnp.where.not(ref_ratio: nil).first
  ReferenceSnp Load (28.7ms)  SELECT  "reference_snps".* FROM "reference_snps" WHERE ("reference_snps"."ref_ratio" IS NOT NULL)  ORDER BY "reference_snps"."id" ASC LIMIT 1
=> #<ReferenceSnp id: 201, rs_id_number: "rs16947", ref_allele: "", alt_allele: "", ref_ratio: #<BigDecimal:7f9e69002238,'0.12E0',9(18)>, alt_ratio: nil, ref_homo: nil, hetero: nil, alt_homo: nil, created_at: "2014-10-29 09:26:13", updated_at: "2015-03-02 02:45:51", gene_id: 117, note: "メモです。\r\nメモです。", chromosome: 22, position: 42523942>
irb(main):008:0> snp.ref_ratio.to_s.encoding 
=> #<Encoding:ASCII-8BIT>
irb(main):009:0> Rails::Html::WhiteListSanitizer.new.sanitize snp.ref_ratio.to_s 
output error : unknown encoding ASCII-8BIT
=> ""
```

The pg gem version is 018.1. What is the root cause?

**Comment by kaspth** [comment]
We're trying to a different approach on #34, can you try that out in your app? :smile:

**Comment by JunichiIto** [comment]
@kaspth I confirmed https://github.com/rails/rails-html-sanitizer/pull/34 fixed the problem in my app. Thanks :smiley:

**Comment by kaspth** [comment]
@JunichiIto thanks, it's merged and we'll ship a version shortly :grin:

**Pull request closed without merging**